proposals/NNNN-task-cancellation-shields.md

```diff
@@ -0,0 +1,340 @@
+* Implementation: [PR #85637](https://github.com/swiftlang/swift/pull/85637)
+* Review: 
+  * TODO
+
+## Introduction
+
+This proposal introduces a new mechanism to temporarily "ignore" task cancellation, called task cancellation shields.
+
+This can be used to ensure certain pieces of code will execute regardless of the task's cancelled status. A common situation where this is useful is running clean-up code, which must execute regardless of a task's cancellation status.
+
+This proposal dovetails nicely with asynchronous defer statements which were recently introduced in [SE-0493: Support `async` calls in `defer` bodies](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0493-defer-async.md), which are frequently used to express such resource clean-up functionality.
+
+## Motivation
+
+Task cancellation is _final_ and can not be ignored or undone. Once a task has been cancelled, it remains cancelled for the rest of its existence. 
+
+Child tasks are also affected by task cancellation, and cancellation propagates throughout the entire task tree, allowing for efficient and holistic cancelling of entire hierarchies of work, represented as a tree of child tasks.
+
+Today, there is no great way to ignore cancellation, and some pieces of code may therefore by accident not execute to completion. This is especially problematic in clean-up or resource tear-down, where a tear-down method's implementation details might be checking for cancellation, however, we _must_ have this code execute, regardless of the task's cancellation status to properly cleanup some resource, like this:
+
+```swift
+extension Resource { 
+  func cleanup() { // our "cleanup" implementation looks correct...
+    system.performAction(CleanupAction())
+  }
+}
+
+extension SomeSystem { 
+  func performAction(_ action: some SomeAction) { 
+    guard Task.isCancelled else {
+      // oh no! 
+      // If Resource.cleanup calls this while being in a cancelled task,
+      // the action would never be performed!
+      return 
+    }
+    // ... 
+  }
+}
+```
+
+In the above example, while we may have implemented the resource clean-up correctly, we may be unaware of the system only performing actions while the task it is executing in is not cancelled. In order to ensure certain actions execute regardless if called from a cancelled or not cancelled task, we're going to have to "shield" the cleanup code from observing the cancellation status of the calling task.
+
+Today, developers work around this problem by creating unstructured tasks, which creates unnecessary scheduling and may have a performance and even correctness impact on such cleanup code:
+
+```swift
+// WORKAROUND, before cancellation shields were introduced
+func example() async {
+  let resource = makeResource()
+  
+  assert(Task.isCancelled())
+  await Task {
+    assert(!Task.isCancelled())
+    await resource.cleanup() 
+  }.value // break out of task tree, in order to ignore cancellation
+}
```

Comment:
nit: code didn't compile & the first assertion seems a bit confusing
```suggestion
func example() async {
  let resource = makeResource()

  await Task {
    assert(!Task.isCancelled)
    await resource.cleanup()
  }.value // break out of task tree, in order to prevent cleanup from observing cancellation
}
```